### PR TITLE
perf: action & selector-based swapper

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Replace a lot of reacty swapper logic with selectors & actions within the Zustand swapper store.

Input computation is now atomic, and she's real speedy.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

The last architceture piece before we can get https://github.com/shapeshift/web/issues/3468 done.

## Risk

High - everything releting to amount computation has been re
-written (though using the same logic).

useEffects and callbacks have been liberally removed in favour of pure actions, so we need to ensure everything still co
mputes at the right times.

## Testing

A full swapper regression test is needed for this one, with
special attention to the amounts on the input and confirmati
on screens.

### Engineering

👆

### Operations

👆

## Screenshots (if applicable)

N/A

### CODE REVIEW

**Comment by 0xApotheosis** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4138** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4138" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4138?utm_source=stack-comment).

File: src/components/Trade/TradeInput.tsx [hunk 1]
```diff
@@ -181,84 +170,20 @@ export const TradeInput = () => {
   const hasValidSellAmount = bnOrZero(sellAmountCryptoPrecision).gt(0)
 
   const handleInputChange = useCallback(
-    async (action: TradeAmountInputField, amount: string) => {
+    (action: TradeAmountInputField, amount: string) => {
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:173]: 🐐

Comment (0xApotheosis) [reply src/components/Trade/TradeInput.tsx:173]: One beautiful, synchronous, atomic action 🤌

File: src/state/zustand/swapperStore/actions.ts [hunk 2]
```diff
@@ -29,12 +32,16 @@ export const updateTradeAmounts =
...
+    set: SetSwapperStoreAction<SwapperStore>,
+    getState: () => SwapperState,
+  ): SwapperAction['toggleIsExactAllowance'] =>
   () =>
     set(
-      state => {
-        state.isExactAllowance = !state.isExactAllowance
-        return state
+      draft => {
+        const currentState = getState()
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/actions.ts:42]: Any reason to pass immer's `getState` and not use the draft state, which, before being mutated, should effectively be the same?

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+import { TradeAmountInputField } from 'components/Trade/types'
+import { bnOrZero } from 'lib/bignumber/bignumber'
+import { fromBaseUnit, toBaseUnit } from 'lib/math'
+import type { SwapperState } from 'state/zustand/swapperStore/useSwapperStore'
+
+const selectAssetPriceRatio = createSelector(
+  (state: SwapperState) => state.buyAssetFiatRate,
+  (state: SwapperState) => state.sellAssetFiatRate,
+  (buyAssetFiatRate, sellAssetFiatRate) => {
+    return bnOrZero(buyAssetFiatRate).dividedBy(bnOrZero(sellAssetFiatRate)).toString()
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/amountSelectors.ts:11]: `toFixed()` for a more accurate ratio?

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+  },
+)
+
+export const selectSellAmountBeforeFeesBaseUnit = createSelector(
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/amountSelectors.ts:15]: `byAction` vernacular

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+      case TradeAmountInputField.BUY_FIAT:
+      case TradeAmountInputField.SELL_FIAT:
+        return toBaseUnit(bnOrZero(amount).dividedBy(sellAssetFiatRate), sellAssetPrecision)
+      default:
+        return '0'
+    }
+  },
+)
+
+export const selectBuyAmountBeforeFeesBaseUnit = createSelector(
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/amountSelectors.ts:37]: `byAction` vernacular

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+  (state: SwapperState) => state.sellAsset?.precision,
+  (assetPriceRatio, sellAmountBeforeFeesBaseUnit, sellAssetPrecision): string | undefined => {
+    if (!sellAssetPrecision) return undefined
+    return bnOrZero(fromBaseUnit(sellAmountBeforeFeesBaseUnit, sellAssetPrecision))
+      .div(assetPriceRatio)
+      .toString()
+  },
+)
+
+export const selectSellAmountBeforeFeesBuyAssetBaseUnit = createSelector(
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/amountSelectors.ts:71]: Can we invert the selectors deps e.g use the base unit to calculate the precision amount ? Something something loss of precision

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+      case TradeAmountInputField.BUY_FIAT:
+      case TradeAmountInputField.SELL_FIAT:
+        return toBaseUnit(bnOrZero(amount).dividedBy(buyAssetFiatRate), buyAssetPrecision)
+      default:
+        return '0'
+    }
+  },
+)
+
+export const selectSellAmountBeforeFeesBuyAsset = createSelector(
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/amountSelectors.ts:59]: `precision` vernacular?

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+export const selectSellAmountBeforeFeesBuyAssetBaseUnit = createSelector(
+  selectSellAmountBeforeFeesBuyAsset,
+  (state: SwapperState) => state.buyAsset?.precision,
+  (sellAmountBeforeFeesBuyAsset, buyAssetPrecision): string | undefined => {
+    if (!buyAssetPrecision) return undefined
+    return toBaseUnit(sellAmountBeforeFeesBuyAsset, buyAssetPrecision)
+  },
+)
+
+export const selectBuyAmountBeforeFeesBuyAsset = createSelector(
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/amountSelectors.ts:80]: Ditto precision vernacular

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+
+export const selectSellAmountBeforeFeesFiat = createSelector(
+  selectSellAmountBeforeFeesBaseUnit,
+  (state: SwapperState) => state.sellAsset?.precision,
+  (state: SwapperState) => state.sellAssetFiatRate,
+  (sellAmountBeforeFeesBaseUnit, sellAssetPrecision, sellAssetFiatRate): string | undefined => {
+    if (!sellAssetPrecision || !sellAssetFiatRate) return undefined
+    return bnOrZero(fromBaseUnit(sellAmountBeforeFeesBaseUnit, sellAssetPrecision))
+      .times(sellAssetFiatRate)
+      .toFixed(2)
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/amountSelectors.ts:229]: We might want to let consumers handle the `toFixed()`, to keep selectors as atomic as possible WRT data they return

**Review by gomesalexandre — COMMENTED** [review]
Initial scrutiny, untested but conceptually looks a lot better and atomic 🥇.
A few comments mostly WRT naming but not blockers

File: src/components/Trade/TradeInput.tsx [hunk 10]
```diff
@@ -181,84 +170,20 @@ export const TradeInput = () => {
...
-    updateFiatBuyAmount,
-    updateFeeAssetFiatRate,
-    updateFees,
-    updateTrade,
-    updateSelectedSellAssetAccountId,
-    updateSelectedBuyAssetAccountId,
-    updateBuyAssetAccountId,
-    updateSellAssetAccountId,
-  ])
+  const handleToggle = useCallback(() => handleAssetToggle(), [handleAssetToggle])
```
Comment (0xdef1cafe) [inline src/components/Trade/TradeInput.tsx:184]: just an alias/relic now?

File: src/components/Trade/hooks/useFiatRateService.tsx [hunk 11]
```diff
@@ -24,7 +24,7 @@ export const useFiatRateService = () => {
   const [usdRatesArgs, setUsdRatesArgs] = useState<UsdRatesInputArgs>(skipToken)
 
   // Selectors
-  const selectedCurrencyToUsdRate = useAppSelector(selectFiatToUsdRate)
+  const selectedCurrencyToUsdRate = useAppSelector(selectFiatToUsdRate).toString()
```
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useFiatRateService.tsx:27]: stringify in selector perhaps, unless there's a reason for this?

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 12]
```diff
@@ -0,0 +1,394 @@
+import { createSelector } from 'reselect'
+import { TradeAmountInputField } from 'components/Trade/types'
+import { bnOrZero } from 'lib/bignumber/bignumber'
+import { fromBaseUnit, toBaseUnit } from 'lib/math'
+import type { SwapperState } from 'state/zustand/swapperStore/useSwapperStore'
+
+const selectAssetPriceRatio = createSelector(
+  (state: SwapperState) => state.buyAssetFiatRate,
```
Comment (0xdef1cafe) [inline src/state/zustand/swapperStore/amountSelectors.ts:8]: all of these can be pulled out/aliased if you prefer for visual cleanliness but no big deal either way

File: src/state/zustand/swapperStore/testData.ts [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+      estimatedGas: '100000',
+      totalFee: '0.0036',
+    },
+    tradeFeeSource: 'THORChain',
+    buyAssetTradeFeeUsd: '8.5196229516043797526224',
+    sellAssetTradeFeeUsd: '0',
+    networkFeeCryptoHuman: '0.0018',
+    networkFeeCryptoBaseUnit: '1800000000000000',
+  },
+} as SwapperState
```
Comment (0xdef1cafe) [inline src/state/zustand/swapperStore/testData.ts:129]: unnecessary cast?

File: src/state/zustand/swapperStore/useSwapperStore.ts [hunk 14]
```diff
@@ -67,6 +69,9 @@ export const useSwapperStore = (() => {
             set,
             'availableSwappersWithMetadata',
           ),
+          handleAssetToggle: handleAssetToggle(set, getState),
+          updateSelectedCurrencyToUsdRate: createUpdateAction(set, 'selectedCurrencyToUsdRate'),
+          handleInputAmountChange: handleInputAmountChange(set, getState),
```
Comment (0xdef1cafe) [inline src/state/zustand/swapperStore/useSwapperStore.ts:74]: can we get a tl;dr on the rationale of dep injecting the `getState` accessor fn here?

File: src/state/zustand/swapperStore/actions.ts [hunk 15]
```diff
@@ -44,19 +51,91 @@ export const clearAmounts =
...
+        draft.trade = undefined
+        draft.buyAmountFiat = '0'
+        draft.sellAmountFiat = '0'
+        return draft
       },
       false,
       `swapper/clearAmounts`,
     )
+
+export const handleAssetToggle =
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/actions.ts]: Fully agree, great suggestion. Renamed in https://github.com/shapeshift/web/pull/4138/commits/22f6923263d9edfca9ea6d42e5082f14ea0a400b

File: src/state/zustand/swapperStore/actions.ts [hunk 16]
```diff
@@ -44,19 +51,91 @@ export const clearAmounts =
...
+    set: SetSwapperStoreAction<SwapperStore>,
+    getState: () => SwapperState,
+  ): SwapperAction['handleInputAmountChange'] =>
+  () =>
+    set(
+      draft => {
+        const currentState = getState()
+        const tradeAmountsByActionAndAmount = selectTradeAmountsByActionAndAmount(currentState)
+        const currentSellAsset = draft.sellAsset
+        const currentBuyAsset = draft.buyAsset
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/actions.ts]: I'm down: https://github.com/shapeshift/web/pull/4138/commits/1a19e737735ec7cad27c0e90d91e0fcc14207446

File: src/state/zustand/swapperStore/amountSelectors.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts]: https://github.com/shapeshift/web/pull/4138/commits/1a19e737735ec7cad27c0e90d91e0fcc14207446

File: src/state/zustand/swapperStore/amountSelectors.ts (hunk 5 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts:38]: https://github.com/shapeshift/web/pull/4138/commits/1a19e737735ec7cad27c0e90d91e0fcc14207446

File: src/state/zustand/swapperStore/amountSelectors.test.ts [hunk 17]
```diff
@@ -0,0 +1,58 @@
+import { TradeAmountInputField } from 'components/Trade/types'
+import { selectTradeAmountsByActionAndAmount } from 'state/zustand/swapperStore/amountSelectors'
+import { baseSwapperState } from 'state/zustand/swapperStore/testData'
+
+describe('calculateAmounts', () => {
+  it('returns cryptoSellAmount, cryptoBuyAmount, fiatSellAmount, fiatBuyAmount for SELL_CRYPTO action', () => {
+    const action = TradeAmountInputField.SELL_CRYPTO
+    const amount = '0.001'
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.test.ts:9]: https://github.com/shapeshift/web/pull/4138/commits/1a19e737735ec7cad27c0e90d91e0fcc14207446

File: src/state/zustand/swapperStore/amountSelectors.ts (hunk 7 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts]: https://github.com/shapeshift/web/pull/4138/commits/1a19e737735ec7cad27c0e90d91e0fcc14207446

File: src/state/zustand/swapperStore/amountSelectors.ts (hunk 8 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts]: https://github.com/shapeshift/web/pull/4138/commits/1a19e737735ec7cad27c0e90d91e0fcc14207446

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+  selectTotalTradeFeeBuyAssetBaseUnit,
+  (sellAmountBeforeFeesBuyAssetBaseUnit, totalTradeFeeBuyAssetBaseUnit): string | undefined => {
+    if (!sellAmountBeforeFeesBuyAssetBaseUnit || !totalTradeFeeBuyAssetBaseUnit) return undefined
+    return bnOrZero(sellAmountBeforeFeesBuyAssetBaseUnit)
+      .minus(totalTradeFeeBuyAssetBaseUnit)
+      .toString()
+  },
+)
+
+export const selectBuyAmountAfterFees = createSelector(
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts]: https://github.com/shapeshift/web/pull/4138/commits/1a19e737735ec7cad27c0e90d91e0fcc14207446

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+  (state: SwapperState) => state.buyAssetFiatRate,
+  (buyAmountAfterFeesBaseUnit, buyAssetPrecision, buyAssetFiatRate): string | undefined => {
+    if (!buyAssetPrecision || !buyAssetFiatRate || !buyAmountAfterFeesBaseUnit) return undefined
+    return bnOrZero(fromBaseUnit(buyAmountAfterFeesBaseUnit, buyAssetPrecision))
+      .times(buyAssetFiatRate)
+      .toFixed(2)
+  },
+)
+
+export const selectAmountBeforeFeesBuyAsset = createSelector(
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts]: https://github.com/shapeshift/web/pull/4138/commits/1a19e737735ec7cad27c0e90d91e0fcc14207446

File: src/components/Trade/hooks/useFiatRateService.tsx (hunk 11 above)
Comment (0xApotheosis) [reply src/components/Trade/hooks/useFiatRateService.tsx]: No reason, good catch: addressed in https://github.com/shapeshift/web/pull/4138/commits/dddbde80b84b26edfea966547554a69f2d10f290.

File: src/components/Trade/TradeInput.tsx (hunk 10 above)
Comment (0xApotheosis) [reply src/components/Trade/TradeInput.tsx]: Indeed, addressed in https://github.com/shapeshift/web/pull/4138/commits/dddbde80b84b26edfea966547554a69f2d10f290.

File: src/state/zustand/swapperStore/amountSelectors.ts (hunk 6 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts]: Absolutely. Inverted in https://github.com/shapeshift/web/pull/4138/commits/dddbde80b84b26edfea966547554a69f2d10f290.

File: src/state/zustand/swapperStore/amountSelectors.ts (hunk 9 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts]: Yes, this is definitely the right way to handle this. Updated all selectors to use full precision in https://github.com/shapeshift/web/pull/4138/commits/10b9d8bcefe0a0c36807e69f55d674898f971837, leaving the 2-decimal rounding to the UI layer.

File: src/state/zustand/swapperStore/testData.ts (hunk 13 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/testData.ts]: Ish - it was to avoid mocking the swapper store actions.

Mocked and removed cast in https://github.com/shapeshift/web/pull/4138/commits/5179db979e52c13eb6e7ed0b5222a0d6aa67993d.

File: src/state/zustand/swapperStore/amountSelectors.ts (hunk 12 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts]: Added convenience selectors in https://github.com/shapeshift/web/pull/4138/commits/f30f3dffcecdcd3834a422c5cfff66ef084220a5

File: src/components/Trade/TradeInput.tsx [hunk 20]
```diff
@@ -62,56 +77,48 @@ const moduleLogger = logger.child({ namespace: ['TradeInput'] })
...
+  const buyAsset = useSwapperStore(selectBuyAsset)
+  const sellAsset = useSwapperStore(selectSellAsset)
+  const buyAmountCryptoPrecision = useSwapperStore(selectBuyAmountCryptoPrecision)
+  const sellAmountCryptoPrecision = useSwapperStore(selectSellAmountCryptoPrecision)
   const updateSellAmountCryptoPrecision = useSwapperStore(
     state => state.updateSellAmountCryptoPrecision,
   )
   const swapperSupportsCrossAccountTrade = useSwapperStore(selectSwapperSupportsCrossAccountTrade)
   const checkApprovalNeeded = useSwapperStore(selectCheckApprovalNeededForWallet)
+  const handleSwitchAssets = useSwapperStore(state => state.handleSwitchAssets)
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:114]: 💜

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.tsx [hunk 21]
```diff
@@ -1,104 +1,22 @@
...
-        updateSellAssetAccountId(undefined)
-        updateBuyAssetFiatRate(undefined)
-        updateSellAssetFiatRate(undefined)
-        updateFeeAssetFiatRate(undefined)
-        updateFees(undefined)
-      }
-
-      clearAmounts()
+    (asset: Asset, action: AssetClickAction) => {
+      handleAssetSelection({ asset, action })
```
Comment (gomesalexandre) [inline src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.tsx:16]: 🤌

File: src/state/zustand/swapperStore/actions.ts [hunk 22]
```diff
@@ -1,62 +1,175 @@
...
+          fiatBuyAmount,
+        } = tradeAmountsByActionAndAmount
+        const buyAmountCryptoPrecision =
+          buyAmountBuyAssetBaseUnit && buyAsset
+            ? fromBaseUnit(buyAmountBuyAssetBaseUnit, buyAsset.precision)
+            : '0'
+        const sellAmountCryptoPrecision =
+          sellAmountSellAssetBaseUnit && sellAsset
+            ? fromBaseUnit(sellAmountSellAssetBaseUnit, sellAsset.precision)
+            : '0'
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/actions.ts:96]: Can `bnOrZero()` + `pow pow` it and move the short circuiting inside `bnOrZero()` if you feel like it

File: src/state/zustand/swapperStore/amountSelectors.test.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/state/zustand/swapperStore/amountSelectors.test.ts:9]: She's still here, if you feel like a global LSP rename of `amount` in `SwapperState` @0xApotheosis

File: src/state/zustand/swapperStore/amountSelectors.ts [hunk 23]
```diff
@@ -0,0 +1,357 @@
...
+  selectSellAsset,
+  selectSellAssetFiatRate,
+} from 'state/zustand/swapperStore/selectors'
+import type { SwapperState } from 'state/zustand/swapperStore/types'
+
+const selectAssetPriceRatio = createSelector(
+  selectBuyAssetFiatRate,
+  selectSellAssetFiatRate,
+  (buyAssetFiatRate, sellAssetFiatRate) => {
+    return bnOrZero(buyAssetFiatRate).dividedBy(bnOrZero(sellAssetFiatRate)).toFixed()
```
Comment (gomesalexandre) [inline src/state/zustand/swapperStore/amountSelectors.ts:21]: Can ditch the inner `bnOrZero()` here since we're in the context of a BN method

**Review by gomesalexandre — COMMENTED** [review]
First code pass, conceptually looks turbo sane 🐐 going to give this guy a proper testing

**Review by gomesalexandre — COMMENTED** [review]
First pass, not getting rugged by losing my draft review this time 😄. Tested:

### ZRX Avalanche

#### Preview

image

#### Approval

image

#### Confirm - amounts are stable after the approval

image

#### Status

image

### THOR swaps

#### Preview - minimum RUNE amount for a refund not to occur is properly handled

image

#### Confirm

image

#### Status

image

#### Tx

image
image

- Not sure if the PR is the right place to handle this, but this is most definitely wrong. "Before fees" is correct, but "After fees" (i.e expected amount) should be "Before fees" minus "After fees", as can be seen on the viewblock Tx.

This is data we *actually* have, since we're sending a minimum amount in the memo.

**Review by gomesalexandre — COMMENTED** [review]
Tested:

### Mainnet ZRX swaps

#### Preview

image
image
image

- Changing the currency in settings while in the trade modal doesn't update the fiat rate as spotted previously, though that's not a concern of this PR

#### Confirm

- Amounts are stable

image

Came back to the input step as a sanity check, then updated the amounts:

image

#### Confirm

- Amounts are matching the newly input amounts, and are stable

image

#### Status

image
image

### CoW swaps

image
image
image

#### Approval

- Estimated fees are correctly displayed both in crypto and fiat

image

Tried coming back to the input step, inputting new amounts and confirming again, confirmed fees are still displayed:

image
image

I was however met with this error, which should have been caught at input step:

image

Confirmed with FOX -> ETH we're not properly validating minimum CoW amounts:

image

Coming back to the input step, interestingly, amounts in the ~2 to 5 CRV range seem happy, but 7 actually fail, which doesn't make sense when considering amounts from a hollistic perspective e.g if 2-5 CRV is valid, 7 should also be

image

Continued with FOX -> ETH to get a happy CoW quote, however, it seems like we may be having the same fee issue as spotted in the previous review:

image
image
image

That doesn't seem like a regression per se, as prod is sending similar requests and allowing similar amounts (actually not deducting fees as it should, so this PR improves things in that regard):

image

Tried with higher sell amount this time around:

image
image

- Minimum amount isn't matching the "You Get" amount in preview step

image

- On-chain order minimum isn't meeting the "Min. after slippage" nor the "You get" at preview step

image

File: src/state/zustand/swapperStore/actions.ts (hunk 22 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/actions.ts:96]: You no like `fromBaseUnit()`?

File: src/state/zustand/swapperStore/amountSelectors.ts (hunk 23 above)
Comment (0xApotheosis) [reply src/state/zustand/swapperStore/amountSelectors.ts]: https://github.com/shapeshift/web/pull/4138/commits/4055a9f837db9ae2527b0b3e36fc8fb4aacb8cc9

**Comment by 0xApotheosis** [comment]
Alright big @gomesalexandre, a summary of what I found from your excellent review:

✅ The incorrect receive amount on the trade confirm screen has been fixed (we were calling an incorrect selector from that component)

🗒️ The receive amount shown does indeed differ from what is received (slightly). This is a hangover from the V1 swapper, which we have yet to fix, as whilst we have the exact amount available on the quote, we instead show a receive amount converted from the buy amount using the fiat rate, through which there is some loss of precision due to rounding an possibly a slight difference in fiat rate vs quote rates. This is currently (I'm pretty sure) the case in prod too. I'll follow up with a PR that uses the actual quote rates, but this doesn't belong in this PR as it's a decent logic change from what we do today.

🗒️ There is something funny going on with the CowSwap fees and expected amount, which I could also replicate in production (e.g. 7 is too low, 3 is fine etc). I'll add another PR up the stack to fix this before we turn on swapper selection, though it should not block this PR.

In summary, I've fixed the known regressions in this PR, and the remaining issues are already in prod - so if you are happy, we should get this in.

File: src/state/zustand/swapperStore/actions.ts (hunk 22 above)
Comment (gomesalexandre) [reply src/state/zustand/swapperStore/actions.ts:96]: Potato potato, just looked like that may oneline better 🍠

**Review by gomesalexandre — APPROVED** [review]
Retested:

### FOX -> ETH ZRX swap

image
image

- "You get"/Min. expected after slippage in input step and and "Min. expected after slippage" in confirm step are matching

### ERC20 <-> ERC20 ZRX swap

#### Input

image

#### Confirm

image

- Amounts are stable 🤌 

#### Status

image
image

#### Tx

image

### CoW FOX -> ETH

#### Input

image

#### Confirm - amounts are stable
image

#### Status

image
image

- Amounts are now matching the on-chain order (minus something something precision, and off-by-x as we're guesstimating the receive amount from the buy amount fiat rate as you've noted)


in
